### Analyzing the C. bovinus with Black data ###

#Loading Libraries
```{R}```
library('devtools')
library('pegas')
library('adegenet')
library('vcfR') 
library('rospca')
library('dartR')
library('zvau')
library('geosphere')
library('stringr')
library('ggmap')
library('ggcompoplot')
library('vegan')
library('spdep')
library('adespatial')
library('igraph')
library('poppr') 
library('smatr')
library('radiator')
library('related')
library('dartR')
library('scales')
library('akima')
library('spacemakeR')
library('hierfstat')
library('ggpubr')
library('PopGenReport')
library('sdmpredictors') 
library('ggord')


#Add functions
```{R}```
#Pulls the unique loci names from alleles selected from genind object names (e.g. dDocent_Contig_23960.001)
Loci_names <- function(NAMES, SEP="[.]", REMOVE=1){
  COL <- length(strsplit(head(NAMES,n=1), SEP)[[1]])
  TMP_U <- unique(matrix(unlist(strsplit(NAMES,SEP)),ncol=COL,byrow=T)[,1:(COL-REMOVE)])
  if(is.matrix(TMP_U)){TMP_DF <- data.frame(TMP_U)
  } else {TMP_DF <- data.frame(matrix(TMP_U, ncol=(COL-REMOVE), byrow=T))}
  return(tidyr::unite(TMP_DF, "loci", 1:ncol(TMP_DF), sep=SEP))
}

#Removing loci for genind object from a vcf; Needs to have "dDocent_Contig_#_Pos" format
#Could add an option to select the locus list or remove it
VCF_remove <- function(genind.vcf, loci.list){
  tmp.loci <- matrix(unlist(strsplit(locNames(genind.vcf),"_")), ncol=4, byrow=T)
  if(sum(is.na(tail(tmp.loci, n=1)))){print("locNames(genind) in incorrect format"); break}
  tmp.loci <- data.frame(Locus=paste(tmp.loci[,1], tmp.loci[,2], tmp.loci[,3],sep="_"), Pos=tmp.loci[,4])

  tmp <- vector()
  for(i in loci.list){
    ROW <- which(tmp.loci$Locus == i)
    tmp <- append(tmp,paste(tmp.loci$Locus, tmp.loci$Pos, sep="_")[ROW])
  }

  set.loc <- subset(locNames(genind.vcf), !locNames(genind.vcf) %in% tmp)
  return(genind.vcf[ ,loc=set.loc])
}

#Importing Data
```{R}```
strata <- read.csv(file = "SNP.TRS.F07.2023-05-16/indv.csv", header = TRUE)
strata$POP2 <- strata$POP
strata$POP2[strata$POP2 %in% c("Cyb-FLS","Cyb-HEAD")] <- "Cyb-DY"

vcf <- read.vcfR(file="SNP.TRS.F07.vcf")
gen.vcf<-vcfR2genind(vcf)
strata(gen.vcf) <- strata[match(indNames(gen.vcf),strata$INDV),]
head(strata(gen.vcf))
rm(vcf)

gen <- read.genepop(file = "SNP.TRS.F07.gen", ncode=3L, quiet = FALSE)
strata(gen) <- strata[match(indNames(gen),strata$INDV),]
head(strata(gen))

#Adding related ID to strata
gen.indv <- substr(gen@strata$INDV, 1, 20)
gen.indv <- data.frame(cbind(as.character(gen@strata$INDV), gen.indv))
names(gen.indv) <- c("gen.ID", "relate.ID")

gen@strata$relate.ID <- gen.indv[match(indNames(gen),gen.indv$gen.ID),"relate.ID"]
head(strata(gen))

#Relatedness within site
```{R}```
min.list <- max.list <- NULL

setPop(gen) <- ~POP
gen.list <- seppop(gen, drop=T)
par(mfrow=c(2,3))

for(i in names(gen.list)){
  tmp.gen <- gen.list[[i]]
  if(nInd(tmp.gen) < 5){next}
  gl <- gi2gl(tmp.gen)
  test.cov <- gl2related(gl, save=F)
  tmp.rel <- coancestry(test.cov, wang =1)$relatedness
  min.list <- c(min.list, min(tmp.rel$wang))
  max.list <- c(max.list, max(tmp.rel$wang))
  hist(tmp.rel$wang, breaks=50, col="red4", main=i)
  assign(paste(i,"rel",sep="_"), tmp.rel)
}

MIN <- min(min.list)
MAX <- max(max.list)

tmp.break <- seq(MIN,MAX,(MAX-MIN)/100)

rel.list <- NULL

png("Site_specific_relatedness.png", res=200, width=8000, height=6000)
par(mfrow=c(2,3))
for(i in names(gen.list)){
  tmp.gen <- gen.list[[i]]
  if(nInd(tmp.gen) < 10){next}
  tmp.rel <- get(paste(i,"rel",sep="_"))
  rel.list <- rbind(rel.list, tmp.rel[tmp.rel$wang > 0.4,c(2,3,6)])
  hist(tmp.rel$wang, breaks=tmp.break, col="red4", main=i, xlim=c(MIN,MAX), xlab="relatedness")
}
dev.off()

#Removing duplicates and related individuals
```{R}```	
rm.list <- list()

for(i in 1:nrow(rel.list)){
  #Only removing duplicates
  if(rel.list$wang[i] < 0.8){next}

  j <- as.character(rel.list[i,])

  tmp <- gen@strata[gen@strata$relate.ID %in% j, ]
  tmp$Diff <- abs(as.numeric(as.character(tmp[ , "E.HOM."])) - as.numeric(as.character(tmp[ , "O.HOM."])))

  Miss.tab <- table(as.numeric(as.character(tmp$N_MISS)))
  Depth.tab <- table(as.numeric(as.character(tmp$MEAN_DEPTH)))

  MIN.Miss <- min(as.numeric(names(Miss.tab)))
  MAX.Depth <- max(as.numeric(names(Depth.tab)))

  if(Miss.tab[names(Miss.tab)==MIN.Miss]==1 && Depth.tab[names(Depth.tab)==MAX.Depth]==1){keep <- as.character(tmp[tmp$N_MISS==MIN.Miss,1])
  } else {keep <- as.character(tmp[tmp$Diff == min(tmp$Diff),1])}
  print(keep)

  rm.list[[i]] <- as.character(tmp[grep(keep, tmp$INDV, invert=T),1])
  rm(keep, tmp)
}

set.ind <- subset(indNames(gen), !indNames(gen) %in% rm.list)
gen2 <- gen[set.ind, ]
set.ind <- subset(indNames(gen.vcf), !indNames(gen.vcf) %in% rm.list)
gen2.vcf <- gen.vcf[set.ind, ]

save(gen2, file="gen2.gz", commpress=T)
#load("gen2.gz")
save(gen2.vcf, file="gen2.vcf.gz", commpress=T)
#load("gen2.vcf.gz")

#Outlier Detection (Outflank)
```{R}```
setPop(gen2.vcf) <- ~POP
tmp <- gl.outflank(gen2.vcf, qthreshold = 0.1)
length(which(tmp$outflank$results[15]=="TRUE" & tmp$outflank$results[11]=="goodH"))

outliers <- tmp$outflank$results[which(tmp$outflank$results[15]=="TRUE" & tmp$outflank$results[11]=="goodH"),1]
out.list <- matrix(unlist(strsplit(as.vector(outliers),"[.]")),ncol=2,byrow=T)[,1]
tmp <- matrix(unlist(strsplit(as.vector(outliers),"[_]")),ncol=4,byrow=T)
gen.out.list<-unique(paste(tmp[,1],tmp[,2],tmp[,3],sep="_"))
length(gen.out.list) 

write.table(gen.out.list, "Outflank_outlier.list", quote=F, col.names=F, row.names=F)

#Fst Outliers (Bayescan)
```{R}```
#Exporting data for Bayescan
setPop(gen2) <- ~POP
writeGenPop(gen2, "Black.Final_all.gen", "Temporal C. bovinus data without dups by POP & Time")

#Converting to BS format
```{bash}```
java -jar /usr/local/bin/PGDSpider2-cli.jar -inputfile Black.Final_all.gen -inputformat GENEPOP -outputfile Black.Final_all.BS -outputformat BAYESCAN -spid ~/bin/genepop_to_BS.spid

#Running Bayescan
```{bash}```
mkdir Bayescan
bayescan Black.Final_all.BS -od ./Bayescan -all-trace -threads 15 -thin 100 -nbp 30 -pr_odds 100
 #notepad cleanup

#Analyzing results
```{bash}```
#All sites
head -n2 ../SNP.TRS.Site2_all.gen | tail -n 1 | sed 's/, /\n/g' > contigs
echo "Contigs" | cat - contigs > tmp; mv tmp contigs
paste contigs SNP.TRS.Site2_al_fst.txt | awk 'NR==1{print $0}NR!=1{$2=""; print}' > fst_out.txt

awk 'NR==1{next;} $4 < 0.05{print $0}' fst_out.txt | wc -l
awk 'NR==1{print $0;next;} $4 < 0.05{print $0}' fst_out.txt | less
awk 'NR==1{next;} $4 < 0.05' fst_out.txt | cut -f1 -d" "> Bayescan_Outliers.list

#Gathering Outlier loci
```{R}```
setPop(gen2) <- ~POP

#Outlier Loci
OF.out <- read.table("Outflank_outlier.list", head=F)[,1]
BS.out <- read.table("Bayescan/Bayescan_Outliers.list", head=F)[,1]

out.list <- unique(c(OF.out, BS.out))
length(out.list)

set.loc <- locNames(gen2)[which(!(locNames(gen2) %in% out.list))]
gen3 <- gen2[, loc=set.loc]
gen3.vcf <- VCF_remove(gen2.vcf, gen.out.list)

gen3.out <- gen2[, loc=out.list]
set.loc <- locNames(gen2)[which(!(locNames(gen2) %in% out.list))]
gen3.out.vcf <- VCF_remove(gen2.vcf, set.loc)

#PCA
```{R}```
X <- scaleGen(gen3, NA.method="mean", scale=F)
pca1 <- dudi.pca(X,cent=T,scale=T,scannf=FALSE,nf=5000)
setPop(gen3) <- ~POP

par(mfrow=c(1,1))
tmp.dat <- cbind(gen3@strata, pca1$li[match(gen3@strata$INDV, rownames(pca1$li)),])

p.net_POP <- ggplot() + geom_point(data=tmp.dat, aes(x=Axis1, y=Axis2, fill=POP), pch=21, size=3) + scale_fill_manual(values=funky(5)) + theme_classic() + 
  labs(x=paste("PC1: ",sprintf("%.2f",(pca1$eig[1]/sum(pca1$eig))*100)," %", sep=""), y=paste("PC2: ",sprintf("%.2f",(pca1$eig[2]/sum(pca1$eig))*100)," %", sep="")) +
  guides(fill=guide_legend(title="Sites", override.aes = list(shape = 23))) + 
  ggtitle(paste("Neutral data (n=",length(locNames(gen3))," loci)", sep="")) + 
  stat_ellipse(data=tmp.dat, aes(x=Axis1, y=Axis2, group=POP, col=POP)) + scale_color_manual(values=funky(5)) + guides(color="none")
p.net_POP

ggsave(p.net_POP, file="PCA_gennet_POP_v2.svg", device="svg")

X.out <- scaleGen(gen3.out, NA.method="mean", scale=F)
pca.out <- dudi.pca(X.out,cent=T,scale=T,scannf=FALSE,nf=5000)
setPop(gen3.out) <- ~POP

par(mfrow=c(1,1))
tmp.dat <- cbind(gen3.out@strata, pca.out$li[match(gen3.out@strata$INDV, rownames(pca.out$li)),])

p.out_POP <- ggplot() + geom_point(data=tmp.dat, aes(x=Axis1, y=Axis2, fill=POP), pch=21, size=3) + scale_fill_manual(values=funky(5)) + theme_classic() + 
  labs(x=paste("PC1: ",sprintf("%.2f",(pca.out$eig[1]/sum(pca.out$eig))*100)," %", sep=""), y=paste("PC2: ",sprintf("%.2f",(pca.out$eig[2]/sum(pca.out$eig))*100)," %", sep="")) +
  guides(fill=guide_legend(title="Sites", override.aes = list(shape = 23))) + 
  ggtitle(paste("Outlier data (n=",length(locNames(gen3.out))," loci)", sep="")) + 
  stat_ellipse(data=tmp.dat, aes(x=Axis1, y=Axis2, group=POP, col=POP)) + scale_color_manual(values=funky(5)) + guides(color="none")
p.out_POP

ggsave(p.out_POP, file="PCA_genout_POP_v2.svg", device="svg")

#Exporting for NeEstimator and Arleqiun
```{R}```
setPop(gen3) <- ~POP2
writeGenPop(gen3, "Black.Final_net.gen", "Neutral C. bovinus data without dups by POP")

#Convert to Arlequin format
```{bash}```
java -jar /usr/local/bin/PGDSpider2-cli.jar -inputfile Black.Final_net.gen -inputformat GENEPOP -outputfile Black.Final_net.arp -outputformat Arlequin -spid ~/bin/genepop_to_arlequin_STD.spid

#Exporting for temporal Ne estimates
```{R}```
#USFW samples
set.indv <- gen3@strata$INDV[gen3@strata$POP2 %in% c("Black_USFW","Cyb-USFW")]
gen.tmp <- gen3[set.indv, , drop=T]
setPop(gen.tmp) <- ~POP2
writeGenPop(gen.tmp, "Black.USFW_net.gen", "Refuge C. bovinus data without dups by POP2")

#Wild samples
set.indv <- gen3@strata$INDV[gen3@strata$POP2 %in% c("Black_DY","Cyb-DY")]
gen.tmp <- gen3[set.indv, , drop=T]
setPop(gen.tmp) <- ~POP2
writeGenPop(gen.tmp, "Black.DY_net.gen", "Wild C. bovinus data without dups by POP2")
